### PR TITLE
Migrate the ability to upload and download binlog to cpp

### PR DESCRIPTION
**Pull request by xige-16**
fix issue: https://github.com/milvus-io/milvus/issues/21266
/kind bug

pick pr to 2.2.0 branch: https://github.com/milvus-io/milvus/pull/22382
Signed-off-by: xige-16 <xi.ge@zilliz.com>

### CODE REVIEW

File: internal/core/src/mmap/Column.h [hunk 1]
```diff
@@ -1,14 +1,18 @@
-// Copyright (C) 2019-2020 Zilliz. All rights reserved.
-//
-// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// Licensed to the LF AI & Data foundation under one
```
Comment (yah01) [inline internal/core/src/mmap/Column.h:1]: keep the files structure same with 2.2.0 branch, also the Column type is not for mmap only

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/core/src/common/type_c.h [hunk 2]
```diff
@@ -69,16 +69,6 @@ typedef struct CProto {
...
-    int64_t field_id;
-    const uint8_t* blob;
-    uint64_t blob_size;
-    int64_t row_count;
-    // Set null to disable mmap,
-    // mmap file path will be {mmap_dir_path}/{segment_id}/{field_id}
-    const char* mmap_dir_path;
-} CLoadFieldDataInfo;
-
 typedef struct CLoadDeletedRecordInfo {
```
Comment (jiaoew1991) [inline internal/core/src/common/type_c.h:72]: This struct should also be deleted, right?

File: internal/core/src/index/StringIndexMarisa.cpp [hunk 3]
```diff
@@ -31,11 +31,77 @@ namespace milvus::index {
...
+
+    // fill str_ids_
+    str_ids_.resize(total_num_rows);
+    int64_t offset = 0;
+    for (auto data : field_datas) {
+        auto slice_num = data->get_num_rows();
+        for (size_t i = 0; i < slice_num; ++i) {
+            auto str_id =
+                lookup(*static_cast<const std::string*>(data->RawValue(i)));
+            assert(valid_str_id(str_id));
```
Comment (jiaoew1991) [inline internal/core/src/index/StringIndexMarisa.cpp:94]: Add some error message here?

**Comment by xige-16** [comment]
rerun ut

File: internal/core/src/index/ScalarIndexSort-inl.h [hunk 4]
```diff
@@ -24,22 +24,64 @@
...
+#include "index/Utils.h"
 
 namespace milvus::index {
 
 template <typename T>
-inline ScalarIndexSort<T>::ScalarIndexSort() : is_built_(false), data_() {
+inline ScalarIndexSort<T>::ScalarIndexSort(
+    storage::FileManagerImplPtr file_manager)
+    : is_built_(false), data_() {
+    if (file_manager != nullptr) {
```
Comment (czs007) [inline internal/core/src/index/ScalarIndexSort-inl.h:35]: better to add a AssertInfo

File: internal/core/src/index/StringIndexMarisa.cpp [hunk 5]
```diff
@@ -31,11 +31,77 @@ namespace milvus::index {
...
+
+    // fill str_ids_to_offsets_
+    fill_offsets();
+
+    built_ = true;
+}
+
 void
 StringIndexMarisa::Build(size_t n, const std::string* values) {
     if (built_) {
```
Comment (czs007) [inline internal/core/src/index/StringIndexMarisa.cpp:107]: better to keep same with other ScalarIndex which is just return

File: internal/core/src/index/Utils.cpp [hunk 6]
```diff
@@ -197,4 +187,36 @@ ParseConfigFromIndexParams(
     return config;
 }
 
+void
+AssembleIndexDatas(std::map<std::string, storage::FieldDataPtr>& index_datas) {
+    if (index_datas.find(INDEX_FILE_SLICE_META) != index_datas.end()) {
```
Comment (czs007) [inline internal/core/src/index/Utils.cpp:192]: better to assert index_datas have INDEX_FILE_SLICE_META

File: internal/core/src/indexbuilder/index_c.cpp [hunk 7]
```diff
@@ -21,41 +21,40 @@
 #include "indexbuilder/IndexFactory.h"
 #include "common/type_c.h"
 #include "storage/Types.h"
+#include "indexbuilder/types.h"
+#include "index/Utils.h"
+#include "pb/index_cgo_msg.pb.h"
+#include "storage/Util.h"
 
 CStatus
 CreateIndex(enum CDataType dtype,
```
Comment (czs007) [inline internal/core/src/indexbuilder/index_c.cpp:30]: you can delete this.

**Pull request merged**